Configure Github Actions

maxlambrecht opened the pull request:
This PR adds the config to run the Gradle build with Github Actions in multiple environments (Ubuntu and MacOs, with Java 8, 11 and 16)
Gradle and Lombok dependency were updated to support building with Java 16.
It also removes the travis configuration.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

rturner3 commented on .github/workflows/gradle.yml:
```diff
@@ -0,0 +1,74 @@
+      - uses: actions/upload-artifact@v2
+        with:
+          name: Package
+          path: build/libs
```
Do we need to remove these Gradle files from the cache as a final step in the build?
```
~/.gradle/caches/modules-2/modules-2.lock
~/.gradle/caches/modules-2/gc.properties
```

I noticed a [comment about it](https://docs.github.com/en/actions/guides/building-and-testing-java-with-gradle#caching-dependencies) in the GitHub documentation.

The first file was also removed by the previous Travis build.

maxlambrecht replied:
> Do we need to remove these Gradle files from the cache as a final step in the build?
> 
> ```
> ~/.gradle/caches/modules-2/modules-2.lock
> ~/.gradle/caches/modules-2/gc.properties
> ```
> 
> I noticed a [comment about it](https://docs.github.com/en/actions/guides/building-and-testing-java-with-gradle#caching-dependencies) in the GitHub documentation.
> 
> The first file was also removed by the previous Travis build.

hmm yeah, we would probably need to delete those. I'll add that step.

rturner3 commented on .github/workflows/gradle.yml:
```diff
@@ -0,0 +1,74 @@
+          java-version: '11'
+          distribution: 'adopt'
+      - name: Build with Gradle
+        run: ./gradlew build
```
It would be nice to also run tests for each JDK + OS combination.

maxlambrecht replied:
Which JDK versions would you consider relevant?

rturner3 replied:
Whichever ones we claim support for - today I believe that's 8, 11, and 16?

rturner3 replied:
Whichever ones are actively supported by Oracle tends to be a reasonable set of versions to maintain compatibility for. Today that would be 8 (LTS), 11 (LTS), and 16 - the same ones you have here.

I now recall that the Gradle `build` target also runs tests, so what you have here looks good!

rturner3 reviewed (commented):
Do we have any way to collect code coverage statistics in the GitHub actions CI builds? I saw that it used to be set up for Travis.

maxlambrecht commented:
> Do we have any way to collect code coverage statistics in the GitHub actions CI builds? I saw that it used to be set up for Travis.

That's a very good point! I didn't look into it, I'll see if I can add that as part of this PR.

maxlambrecht commented:
> > Do we have any way to collect code coverage statistics in the GitHub actions CI builds? I saw that it used to be set up for Travis.
> 
> That's a very good point! I didn't look into it, I'll see if I can add that as part of this PR.

I addressed all the comments, except I'm still working on the coverage.

maxlambrecht commented:
I configured a `build-and-test` workflow to run on push and pull_request events, using OS/Java combinations (Ubuntu, java 8, 11, 16 /  Macos,  java 8, 11, 14). Java 16 is not installed in Macos.

I also configured a `create-coverage` workflow to run when pushing , it sends the coverage repo to Coveralls, the same way we were doing. It run only on push because it needs a token that is retrieved from a Github Secret and `Secrets are not passed to workflows that are triggered by a pull request from a fork.`

I also updated the CI badge on the README.

rturner3 commented on .github/workflows/gradle.yml:
```diff
@@ -0,0 +1,73 @@
+          rm -f ~/.gradle/caches/modules-2/modules-2.lock
+          rm -f ~/.gradle/caches/modules-2/gc.properties
+
+  builad-and-test-on-macos:
```
```suggestion
  build-and-test-on-macos:
```

rturner3 commented on .github/workflows/gradle.yml:
```diff
@@ -0,0 +1,73 @@
+    runs-on: macos-latest
+    strategy:
+      matrix:
+        java-version: [ 8, 11, 14] # macos-latest at his point doesn't have java 16 installed
```
spacing nit
```suggestion
        java-version: [ 8, 11, 14 ] # macos-latest at his point doesn't have java 16 installed
```

rturner3 reviewed (approved):
Looks great, thanks for setting it up!
